specifyweb/specify/datamodel.py

```diff
@@ -8202,7 +8203,147 @@
+            Relationship(name='modifiedByAgent', type='many-to-one', required=False, relatedModelName='Agent', column='ModifiedByAgentID'),
+        ],
+        fieldAliases=[
+
+        ]
+    ),
+    Table( # CollectionObjectGroupJoin
+        sp7_only=True,
+        django_app='specify',
+        classname='edu.ku.brc.specify.datamodel.CollectionObjectGroupJoin',
+        table='collectionobjectgroupjoin',
+        tableId=1017,
+        idColumn='CollectionObjectGroupJoinID',
+        idFieldName='collectionObjectGroupJoinId',
+        idField=IdField(name='collectionObjectGroupJoinId', column='CollectionObjectGroupJoinID', type='java.lang.Integer'),
+        fields=[
+            Field(name='isprimary', column='IsPrimary', indexed=False, unique=False, required=True, type='java.lang.Boolean'),
+            Field(name='issubstrate', column='IsSubstrate', indexed=False, unique=False, required=True, type='java.lang.Boolean'),
+            Field(name='precedence', column='Precedence', indexed=False, unique=False, required=True, type='java.lang.Integer'),
+            Field(name='version', column='Version', indexed=False, unique=False, required=False, type='java.lang.Integer'),
+            Field(name='timestampcreated', column='TimestampCreated', indexed=False, unique=False, required=True, type='java.sql.Timestamp'),
+            Field(name='timestampmodified', column='TimestampModified', indexed=False, unique=False, required=False, type='java.sql.Timestamp'),
+            Field(name='text1', column='Text1', indexed=False, unique=False, required=False, type='java.lang.String', length=255),
+            Field(name='text2', column='Text2', indexed=False, unique=False, required=False, type='java.lang.String', length=255),
+            Field(name='text3', column='Text3', indexed=False, unique=False, required=False, type='java.lang.String', length=255),
+            Field(name='integer1', column='Integer1', indexed=False, unique=False, required=False, type='java.lang.Integer'),
+            Field(name='integer2', column='Integer2', indexed=False, unique=False, required=False, type='java.lang.Integer'),
+            Field(name='integer3', column='Integer3', indexed=False, unique=False, required=False, type='java.lang.Integer'),
+            Field(name='boolean1', column='Boolean1', indexed=False, unique=False, required=False, type='java.lang.Boolean'),
+            Field(name='boolean2', column='Boolean2', indexed=False, unique=False, required=False, type='java.lang.Boolean'),
+            Field(name='boolean3', column='Boolean3', indexed=False, unique=False, required=False, type='java.lang.Boolean')
+        ],
+        indexes=[
+
+        ],
+        relationships=[
+            Relationship(name='parentcog', type='many-to-one', required=True, relatedModelName='CollectionObjectGroup', column='ParentCOGID', dependent=True),
```

Comment:
from 

> I think the link from CoJo to COG should be independent (as COGs exist with our without the COJO record), but a COJO record can't exist without a COG and CO or COG and COG.

Isn't what we currently have the reverse of that?

Reply:
I don't think we need to have explict dependent. just make this not nullable and call it done..

Reply:
"Dependent" controls the behavior of the app (changes how WB, QCBX/subviews, deletion, etc. behave) so it is important to configure it appropriately. 

If we make it not nullable, is the default is independent?

Reply:
> "Dependent" controls the behavior of the app (changes how WB, QCBX/subviews, deletion, etc. behave) so it is important to configure it appropriately.

I know 😉 

You don't understand what I want to bring up. It should be on the otherside, not here. 

See https://github.com/specify/specify7/blob/1143595eef287dd6523826ad721c7885e4a767b2/specifyweb/specify/datamodel.py#L47-L48 
for a baby example

Reply:
Make it non-nullable will make it dependent at schema level. 

Make it dependent on the otherside (COG table) will make it so that all the API retrievals fetch dependent.

And, you are wrong about workbench. It doesn't care if anything is dependent or not. It cares about to-many and NOT DEPENDENT. Don't just assume how things work

Reply:
The way we set it up in the schema (DB) changes how things can be uploaded in the WorkBench. That is inarguable. If something requires a FK to exist (i.e. Determination to CO), it can't be uploaded since it is dependent on a CO. This isn't due to the nature of Specify's data model dependencies, but it is due to a database dependency.

I'm not arguing against your suggestion, just stating something we should keep in mind. My goal is to represent the interests of our users and to follow the conclusions of the GS analysts meetings.

Please be mindful of your tone and make sure that you are respectful to me and other team members in both your review comments and replies.

I do appreciate your insight and expertise in this area, so please continue to provide feedback as we develop further schema changes!

Reply:
> That is inarguable. If something requires a FK to exist (i.e. Determination to CO), it can't be uploaded since it is dependent on a CO

Okay, let's not mix things up. By dependent, I mean in the datamodel. Not in the database. For example, if you were to do a workbench upload without `dependent=True` on the current geo schema, with COJO as the base table and had COG empty, it will raise an error. Why? Bc it is `not null`. See below:
https://github.com/specify/specify7/blob/b40008b7c26b51369f16eabc91cf1e9db102ff6a/specifyweb/specify/migrations/0002_geo.py#L154
Notice that's the migration file, the absolute "truth".

So, again, dependent (in the datamodel) really doesn't do anything as far as workbench is concerned. Yes, you cannot upload determination without co. Is that bc it is dependent? No. It is because relationship to CO is `not null`

https://github.com/specify/specify7/blob/b40008b7c26b51369f16eabc91cf1e9db102ff6a/specifyweb/specify/migrations/0001_initial.py#L746

You can absolutely upload records to a dependent table (in the datamodel) without the actual fk. See the below where I was able to make collectionobjectattribute without the corresponding collectionobject, even though collectionobjectattribute is a dependent (in the datamodel) of collectionobject. Am I saying this is correct behaviour? No. Am I explaining what it currently does? Yes.

![image](https://github.com/specify/specify7/assets/61122018/48e9f0ce-32c6-41d1-a301-60c4d537a021)

![image](https://github.com/specify/specify7/assets/61122018/68b4d1d4-6daa-4106-975b-d986cf88a3d9)

Do you see the difference now? 

To make things dependent (at a database level), we just need to have a not null fk (or a constraint check on the otherside - for, say, COA). That's it. To make things more manageable for the user, we need to have concept of "dependent" in the datamodel. They are different from another, entirely. Yes, dependent at schema may not need dependent at datamodel. dependent at datamodel should (must?) need dependent at schema - to simplify logic at code level, and have schema (db) do the heavy integrity checks at we need business rules for.

Reply:
I understand the distinction between the two kinds of "dependent" (in the data model and in the database schema). 

It seems evident that the WorkBench should adhere to our _data model_ dependencies, but I understand this is not the case.

While I appreciate the patronizing explanation, the main point I am making is this:

> a COJO record can't exist without a COG and CO or COG and COG.

However we need to accomplish it, COJO records should only exist if it is joining a CO to a COG or a COG to another COG. Based on the current model, it seems that requirement is met as the record cannot be created due to the relationship to COG being `not null`.

To make things more intuitive for our users, we should ensure the data model has the appropriate relationships set as "dependent". Just like in the case of Collecting Event -> Collectors, those Collectors should not exist without the Collecting Event. We currently indicate that this relationship is "dependent" in the model. 

In my opinion, as COJO is a join table that serves a similar purpose, it's reasonable to set this here as well. When a COG is deleted, I would expect it to be desirable for the associated COJO records should also be removed. It seems this should be handled from the COG side, not the COJO side. When a COJO is removed, the COG should remain.

Since we are managing both the COG parent/child hierarchy and the COG child relationships in this table, it warrants further discussion on how this impacts behavior.

Reply:
Yep, things are more clearer now! 

> It seems this should be handled from the COG side, not the COJO side. When a COJO is removed, the COG should remain.

That's what my original remark was, cryptically, here: https://github.com/specify/specify7/pull/5032#discussion_r1672995137. Setting dependent=True will make COG the dependent of COJO. So, say, when you fetch / view COJO, you'll get COG. What we, from my understanding, need is the reverse. So, COG should have a relationship to COJO (without any fk on COG's side ofc) with `dependent=True`. From the original phrasing of requirement, I thought that we not need dependent in datamodel. I took it too literally.

Managing child/parent is tricky in the same table, from code perspective. Also, @acwhite211, making `('parentcog', 'childcog')` unique can be a bad idea. What if we want to add another collection object that shares the same parentcog and childcog? That'll be rejected from the current design.

Reply:
Good point about the uniqueness constraint!

+1 to Vinny's comment all around ^

Reply:
Maybe we could do something like this, where we create virtual fields on the COG and CO tables pointing to the CoJo table, and making those relationships dependent.  So those fields would only exist in the datamodel, not the database schema.  I'll have to test it out to see if it will work.
https://github.com/specify/specify7/commit/7fe92508e69674876d378afcc9e605124d817777#diff-afd7cb2544aab2415be973bc6a1a994c01d3ebac441e1c2e9c2953901106ea9c
https://github.com/specify/specify7/blob/7fe92508e69674876d378afcc9e605124d817777/specifyweb/specify/datamodel.py#L1596
https://github.com/specify/specify7/blob/7fe92508e69674876d378afcc9e605124d817777/specifyweb/specify/datamodel.py#L8276-L8277

Reply:
Almost. see https://github.com/specify/specify7/pull/5032#discussion_r1674243533, and solution a comment below